File: lua/min_max.lua
function min(a, b)
    return a < b and a or b
end

function max(a, b)
    return a > b and a or b
end

function minMax(a, b)
    return a > b and a, b or b, a
end

function maxElement(seq)
    local max = #seq == 0 and nil or seq[1]
    for i in seq do
        max = max > i and max or i
    end
    return max
end

function minElement(seq)
    local min = #seq == 0 and nil or seq[1]
    for i in seq do
        min = min < i and min or i
    end
    return min
end

function minmaxElement(seq)
    local min = #seq == 0 and nil or seq[1]
    local max = #seq == 0 and nil or seq[1]
    for i in seq do
        min = min < i and min or i
        max = max > i and max or i
    end
    return min, max
end

function lexicographicalCompare(s1, s2)
    if #s1 > #s2 then return false end
    for i = 1, #s1 do
        if s1[i] < s2[i] then
            return true
        elseif s1[i] > s2[i] then
            return false
        end
    end
    return true
end

function isPermutation(s1, s2)
    if #s1 ~= #s2 then return false end
    elements = {}
    for i in s1 do
        elements[i] = elements[i] == nil and 1 or (elements[i] + 1)
    end
    for j in s2 do
        if elements[j] ~= nil and elements[j] > 0 then
            elements[j] = elements[j] - 1
        else
            return false
        end
    end
    return true
end

function nextPermutation(seq)
    if #seq <= 1 then return false end
    local i = #seq
    while true do
        local pos1 = i
        i = i - 1
        if seq[i] < seq[pos1] then
            local pos2 = #seq
            while seq[i] >= seq[pos2] do
                pos2 = pos2 - 1
            end
            seq[i], seq[pos2] = seq[pos2], seq[i]
            local pivot = math.floor((#seq - pos1) / 2)
            for j = 0, pivot do
                seq[#seq-j], seq[pos1+j] = seq[pos1+j], seq[#seq-j]
            end
            return true
        end
        if i == 1 then
            for j in 1, math.floor(#seq / 2) do
                seq[j], seq[#seq-j] = seq[#seq-j], seq[j]
            end
            return false
        end
    end
end

function prevPermutation(seq)
    if #seq <= 1 then return false end
    local i = #seq
    while true do
        local pos1 = i
        i = i - 1
        if seq[i] < seq[pos1] then
            local pos2 = #seq
            while seq[i] >= seq[pos2] do
                pos2 = pos2 - 1
            end
            seq[i], seq[pos2] = seq[pos2], seq[i]
            local pivot = math.floor((#seq - pos1) / 2)
            for j = 0, pivot do
                seq[#seq-j], seq[pos1+j] = seq[pos1+j], seq[#seq-j]
            end
            return true
        end
        if i == 1 then
            for j in 1, math.floor(#seq / 2) do
                seq[j], seq[#seq-j] = seq[#seq-j], seq[j]
            end
            return false
        end
    end
end

return {
    min = min,
    max = max,
    minElement = maxElement,
    maxElement = maxElement,
    minMax = minMax,
    minmaxElement = minmaxElement,
    lexicographicalCompare = lexicographicalCompare,
    isPermutation = isPermutation,
    nextPermutation = nextPermutation,
    prevPermutation = prevPermutation
}

File: php/MinMax.php
<?php
function _min($a, $b) {
    return $a < $b ? $a : $b;
}

function _max($a, $b) {
    return $a > $b ? $a : $b;
}

function minMax($a, $b) {
    return $a < $b ? [$a, $b] : [$b, $a];
}

function minElement(array $ary) {
    $reduceMin = function($min, $cur) {
        return (isset($min) && $min < $cur) ? $min : $cur;
    };
    return array_reduce($ary, $reduceMin);
}

function maxElement(array $ary) {
    $reduceMax = function($max, $cur) {
        return $max > $cur ? $max : $cur;
    };
    return array_reduce($ary, $reduceMax);
}

function minMaxElement(array $ary) {
    $reduceMinMax = function($minMax, $cur) {
        list($min, $max) = $minMax;
        $min = (isset($min) && $min < $cur) ? $min : $cur;
        $max = $max > $cur ? $max : $cur;
        return [$min, $max];
    };
    return array_reduce($ary, $reduceMinMax);
}

function lexicographicalCompare($s1, $s2) {
    $len1 = strlen($s1);
    $len2 = strlen($s2);
    $bound = min($len1, $len2);
    for ($i = 0; $i < $bound; $i++) {
        if ($s1[$i] !== $s2[$i]) {
            return $s1[$i] < $s2[$i];
        }
    }
    return $len1 < $len2;
}

function isPermutation(array $ary1, array $ary2) {
    $len1 = count($ary1);
    $len2 = count($ary2);
    if ($len1 != $len2) return false;
    $elems = array();
    foreach($ary1 as $e)
        $elems[$e] = isset($elems[$e]) ? ($elems[$e] + 1) : 1;
    foreach($ary2 as $e) {
        if (empty($elems[$e]))
            return false;
        $elems[$e]--;
    }
    return true;
}

function nextPermutation(array &$ary) {
    $len = count($ary) - 1;
    if ($len <= 0) return false;
    $i = $len;
    while (true) {
        $pos1 = $i;
        $i--;
        if ($ary[$i] < $ary[$pos1]) {
            $pos2 = $len;
            while ($ary[$i] >= $ary[$pos2]) {
                $pos2--;
            }
            $tmp = $ary[$i];
            $ary[$i] = $ary[$pos2];
            $ary[$pos2] = $tmp;
            $pivot = floor(($len+1-$pos1) / 2);
            for ($j = 0; $j < $pivot; $j++) {
                $tmp = $ary[$pos1+$j];
                $ary[$pos1+$j] = $ary[$len-$j];
                $ary[$len-$j] = $tmp;
            }
            return true;
        }
        if ($i === 0) {
            $ary = array_reverse($ary);
            return false;
        }
    }
}

function prevPermutation(array &$ary) {
    $len = count($ary) - 1;
    if ($len <= 0) return false;
    $i = $len;
    while (true) {
        $pos1 = $i;
        $i--;
        if ($ary[$pos1] < $ary[$i]) {
            $pos2 = $len;
            while ($ary[$pos2] >= $ary[$i]) {
                $pos2--;
            }
            $tmp = $ary[$i];
            $ary[$i] = $ary[$pos2];
            $ary[$pos2] = $tmp;
            $pivot = floor(($len+1-$pos1) / 2);
            for ($j = 0; $j < $pivot; $j++) {
                $tmp = $ary[$pos1+$j];
                $ary[$pos1+$j] = $ary[$len-$j];
                $ary[$len-$j] = $tmp;
            }
            return true;
        }
        if ($i === 0) {
            $ary = array_reverse($ary);
            return false;
        }
    }
}
?>
